### PR TITLE
Add fuzz tests for trade flow

### PR DESCRIPTION
**Pull request by manhlx3006**

### CODE REVIEW

File: contracts/sol6/KyberNetwork.sol [hunk 1]
```diff
@@ -1112,12 +1112,16 @@ contract KyberNetwork is WithdrawableNoModifiers, Utils5, IKyberNetwork, Reentra
             }
             totalBps += reservesData.splitsBps[i];
 
-            destAmount += calcDstQty(
+            uint256 destQty = calcDstQty(
```
Comment (ilanDoron) [inline contracts/sol6/KyberNetwork.sol:1115]: this fix probably doesn't belong here.

File: scripts/networkSimulator.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            await reserve.withdrawAllToken(token.address);
+            let initialTokenAmount = new BN(2000000).mul(new BN(10).pow(new BN(await token.decimals())));
+            await token.transfer(reserve.address, initialTokenAmount);
+            await reserve.setRate(token.address, tokensPerEther, ethersPerToken);
+        }
+        ethSender++;
+    }
+}
+
+async function zeroNetworkBalance(network, tokens, admin) {
```
Comment (ilanDoron) [inline scripts/networkSimulator.js:255]: better move this function to helper
so we don't duplicate code.

File: contracts/sol6/KyberNetwork.sol (hunk 1 above)
Comment (manhlx3006) [reply contracts/sol6/KyberNetwork.sol:1115]: yes, already added in description of this PR, some fixes here just to be able running the fuzz tests

File: scripts/networkSimulator.js (hunk 2 above)
Comment (manhlx3006) [reply scripts/networkSimulator.js:255]: Ok

**Comment by manhlx3006** [comment]
Wrapped networkSimulator into a function that is triggered from tradeFuzzTests.js file. Can update numberLoops if want to run more trades

File: test/sol6/tradeFuzzTests.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        await nwHelper.addReservesToStorage(storage, reserveInstances, tokens, operator);
+    });
+
+    it(`Run trade simulator with ${numberLoops} loops`, async() => {
+        await networkSimulator.doFuzzTradeTests(
+            network, networkProxy, storage, matchingEngine,
+            reserveInstances, accounts, tokens, numberLoops
+        );
+    });
+});
```
Comment (Anyhowclick) [inline test/sol6/tradeFuzzTests.js:152]: CRLF =)

Comment (manhlx3006) [reply test/sol6/tradeFuzzTests.js]: Fixed =)))

File: contracts/sol6/KyberNetwork.sol (hunk 1 above)
Comment (manhlx3006) [inline contracts/sol6/KyberNetwork.sol:1115]: Fixed created: https://github.com/KyberNetwork/smart-contracts/pull/964

File: scripts/networkSimulator.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    }
+}
+
+function logTestResults() {
+    console.log(`--- SUMMARY RESULTS ---`)
+    console.log(`${numberSuccessfulTrades} succesful trades`);
+    console.log(`${numberUpdateReserveRates} times updating reserve rates`);
+    console.log(`${numberGettingsZeroRates} times getting zero for rates`);
+    for(let i = 0; i < listRevertedReasons.length; i++) {
+        console.log(`${numberRevertedTrades[listRevertedReasons[i]]} reverted trades with reason: ${listRevertedReasons[i]}`);
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply scripts/networkSimulator.js:244]: Updated

File: scripts/trades/tradeParamsGenerator.js [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    PLATFORM_FEE_HIGH: "platformFee high", // platform fee >= 10000
+    TOTAL_FEES_HIGH: "fees high", // platform + 2 * network fee >= 10000
+    NETWORK_DISABLED: "network disabled",
+    SENDER_NOT_PROXY: "bad sender", // sender is not network proxy
+    GAS_PRICE_HIGH: "gas price", // gas price is higher than max gas price
+    ZERO_SRC_AMOUNT: "0 srcAmt", // src amount is 0
+    DEST_ADDRESS_ZERO: "dest add 0", // dest address is 0x0
+    SAME_SRC_DEST: "src = dest", // same src and dest token
+    SRC_AMOUNT_HIGH: "srcAmt > MAX_QTY", // src amount > max_qty (10^28)
+
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply scripts/trades/tradeParamsGenerator.js:51]: Added all reverted messages from proxy and network for a trade

File: scripts/trades/tradeParamsGenerator.js [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+        return getRandomInt(0, 100) <= 50 ? MASK_IN_HINTTYPE : MASK_OUT_HINTTYPE;
+    }
+
+    let rand = getRandomInt(0, 100);
+    if (rand < 40) return MASK_IN_HINTTYPE;
+    if (rand < 70) return MASK_OUT_HINTTYPE;
+    return SPLIT_HINTTYPE;
+}
+
+function generateReserveDataWithHintError(hintType, errorType, reserves) {
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply scripts/trades/tradeParamsGenerator.js:574]: Updated, list of selected and splits are now randomized

File: scripts/networkSimulator.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    console.log(`Loop ${loop}: Getting zero rates for trade`);
+    numberGettingsZeroRates++;
+    return false;
+}
+
+// test should revert, there are 4 cases:
+// test is reverted with unspecified, e.g: sender not enough src token, not enough allowance
+// test is reverted but with assertion
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply scripts/networkSimulator.js:163]: I actually don't have any cases that revert will assertion, just put all available options here (all types of revert)

File: scripts/networkSimulator.js [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+                tradeData.maxDestAmount,
+                tradeData.minConversionRate,
+                taker, // platform wallet
+                tradeData.platformFeeBps,
+                tradeData.hint,
+                {from: taker, value: callValue, gasPrice: gasPrice}
+            )
+        )
+    } else if (errorMessage == "assertion") {
+        await expectRevert.assertion(
```
[the remark this answers is not in the document]
Comment (manhlx3006) [reply scripts/networkSimulator.js:184]: I don't have any tests that supposed to assert. But if we can catch some tests that go to this option, it means we should take a look. I will add this to summary message

**Comment by ilanDoron** [comment]
ok

so please for assert make it stand out very clear!!!

File: scripts/trades/multiple_trade_fuzz_test.sh [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+for _ in {1..100}
+do
+    npx buidler test --no-compile test/sol6/tradeFuzzTests.js
+done
```
Comment (Anyhowclick) [inline scripts/trades/multiple_trade_fuzz_test.sh:6]: CRLF

Comment (Anyhowclick) [reply scripts/trades/multiple_trade_fuzz_test.sh:6]: Looking on mobile, I might be wrong

**Pull request merged**